test/parallel/test-dgram-socket-buffer-size.js

```diff
@@ -0,0 +1,94 @@
+'use strict';
+
+const common = require('../common');
+const assert = require('assert');
+const dgram = require('dgram');
+
+{
+  // Should throw error if the socket is never bound.
+  const socket = dgram.createSocket('udp4');
+
+  assert.throws(() => {
+    socket.setRecvBufferSize(8192);
+  }, common.expectsError({
+    code: 'ERR_SOCKET_BUFFER_SIZE',
+    type: Error,
+    message: /^Could not get or set buffer size:.*$/
+  }));
+
+  assert.throws(() => {
+    socket.setSendBufferSize(8192);
+  }, common.expectsError({
+    code: 'ERR_SOCKET_BUFFER_SIZE',
+    type: Error,
+    message: /^Could not get or set buffer size:.*$/
+  }));
+
+  assert.throws(() => {
+    socket.getRecvBufferSize();
+  }, common.expectsError({
+    code: 'ERR_SOCKET_BUFFER_SIZE',
+    type: Error,
+    message: /^Could not get or set buffer size:.*$/
+  }));
+
+  assert.throws(() => {
+    socket.getSendBufferSize();
+  }, common.expectsError({
+    code: 'ERR_SOCKET_BUFFER_SIZE',
+    type: Error,
+    message: /^Could not get or set buffer size:.*$/
+  }));
+}
+
+{
+  // Should throw error if invalid buffer size is specified
+  const socket = dgram.createSocket('udp4');
+
+  socket.bind(common.mustCall(() => {
+    assert.throws(() => {
+      socket.setRecvBufferSize(-1);
+    }, common.expectsError({
+      code: 'ERR_SOCKET_BAD_BUFFER_SIZE',
+      type: Error,
```

Comment:
`common.expectsError()` is subtly broken. It relies on `instanceof` to check the error type. So a `TypeError` is an instance of `Error`, which is why this passed. The simplest fix might be to check `error.constructor` inside of `common.expectsError()`. But that will require updating a bunch of tests because we're now using custom errors all over the place.

cc: @nodejs/testing